### PR TITLE
Semantic Segmentation and base Segmentation module

### LINKED ISSUE #7: Semantic segmentation module
### Semantic Segmentation module

Perform semantic segmentation on aerial images to identify areas of interest in the field.

### PR DESCRIPTION
**Pull request by mirceatlx**
At the moment, the actual model does is not trained so it returns basically random masks. Moreover, it only supports three channels for the input image (not 4 as the images in the Agriculture Vision dataset.)

Closes #7

### CODE REVIEW

**Comment by mirceatlx** [comment]
@paulmis I think the MR is ready

File: src/pipeline/modules/segmentation.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    Args:
+        img: the input image(s)
+        rest: non-specific module arguments
+    """
+    def run(self, data: Data):
+        self.prepare(data)
+        model = keras.models.load_model(self.paths[data.input[0].channels])
+        predictions = []
+        for image in data.modules[Modules.PREPROCESS]["clipping"]:
+            predictions.append(model.predict(np.expand_dims((image.get()), axis=0)))
```
Comment (paulmis) [inline src/pipeline/modules/segmentation.py:34]: Isn't `axis` 0 by default?

File: src/pipeline/modules/segmentation.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    """
+    Perform inference using the images given.
+    Each image should adhere to specific dimensions in order to be
+    used by the model.
+    The model returns a mask for each of the predefined labels that can be
+    used to highlight areas of interest.
+
+    Args:
+        img: the input image(s)
+        rest: non-specific module arguments
```
Comment (paulmis) [inline src/pipeline/modules/segmentation.py:27]: Outdated method args

File: src/pipeline/modules/index/indicies.py [hunk 3]
```diff
@@ -5,4 +5,11 @@
...
-    NUTRIENT = "NUTRIENT"
\ No newline at end of file
+    NUTRIENT = "NUTRIENT",
+    DOUBLEPLANT = "DOUBLEPLANT",
+    DRYDOWN = "DRYDOWN",
+    ENDROW = "ENDROW",
+    PLANTERSKIP = "PLANTERSKIP",
+    WATER = "WATER",
+    WATERWAY = "WATERWAY",
+    WEEDCLUSTER = "WEEDCLUSTER"
```
Comment (paulmis) [inline src/pipeline/modules/index/indicies.py]: Can you add very short descriptions of the indices, like:

```
"Normalized Differential Vegetation Index. Suggests whether the area contains live vegetation"
NDVI = "NDVI"
```

**Comment by paulmis** [comment]
Can you add some documentation akin to what I did in `pipeline/README.md`? I don't think describing intimate details is important, but a high-level description of how segmentation is utilized in the pipeline would be great!

**Comment by paulmis** [comment]
I saw there are 2 bigger files in there (~45MB), are these models/model params?

File: src/pipeline/modules/index/runnables/nutrient.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    Args:
+        masks: masks for each of the pathes
+        patches: cropped parts of fixed size from the mosaic image
+
+    Returns:
+        Nutrient deficit map
+    """
+    def calculate(self, masks, patches) -> list[np.array]:
+        results = [cv2.addWeighted(mask, 1, image, 1, 0, image) for mask, image in zip(masks, patches)]
+        # TODO: concat the results
```
[the remark this answers is not in the document]
Comment (mirceatlx) [reply src/pipeline/modules/index/runnables/nutrient.py:45]: No, I will do it in the next sprint.

File: src/pipeline/modules/segmentation.py (hunk 1 above)
Comment (mirceatlx) [reply src/pipeline/modules/segmentation.py:34]: There is no default argument for the axis parameter.

File: src/pipeline/modules/index/indicies.py (hunk 3 above)
Comment (mirceatlx) [reply src/pipeline/modules/index/indicies.py]: sure

**Comment by mirceatlx** [comment]
Yes, the two bigger files are the params for the models.
